src/main/main.cpp

```diff
@@ -3,529 +3,386 @@
+
+public:
+    Step3Impl(
+        std::string tag,
+        std::shared_ptr<Backend::CassandraBackend> backend)
+        : tag_{std::move(tag)}, backend_{backend}
     {
-        auto const page = doTryFetchLedgerPage(
-            timer, backend, cursor, ledgerRange.minSequence, yield);
-
-        // For each object in page
-        for (auto const& object : page.objects)
-        {
-            auto const objectNFTs = getNFTDataFromObj(
-                ledgerRange.minSequence,
-                std::string(object.key.begin(), object.key.end()),
-                std::string(object.blob.begin(), object.blob.end()));
-            toWrite.insert(toWrite.end(), objectNFTs.begin(), objectNFTs.end());
-        }
-
-        toWrite = maybeDoNFTWrite(toWrite, backend, stepTag);
-        cursor = page.cursor;
-    } while (cursor.has_value());
-
-    doNFTWrite(toWrite, backend, stepTag);
-}
-
-static void
-doMigrationStepThree(Backend::CassandraBackend& backend)
-{
-    /*
-     * Step 3 - Drop the old `issuer_nf_tokens` table, which is replaced by
-     * `issuer_nf_tokens_v2`. Normally, we should probably not drop old tables
-     * in migrations, but here it is safe since the old table wasn't yet being
-     * used to serve any data anyway.
-     */
-    std::stringstream query;
-    query << "DROP TABLE " << backend.tablePrefix() << "issuer_nf_tokens";
-    CassStatement* issuerDropTableQuery =
-        cass_statement_new(query.str().c_str(), 0);
-    CassFuture* fut =
-        cass_session_execute(backend.cautionGetSession(), issuerDropTableQuery);
-    CassError const rc = cass_future_error_code(fut);
-    cass_future_free(fut);
-    cass_statement_free(issuerDropTableQuery);
-    backend.sync();
-
-    if (rc != CASS_OK)
-        clio::LogService::warn() << "Could not drop old issuer_nf_tokens "
-                                    "table. If it still exists, "
-                                    "you should drop it yourself\n";
-}
+    }
 
-static void
-doMigration(
-    clio::Config const& config,
-    Backend::CassandraBackend& backend,
-    boost::asio::steady_timer& timer,
-    boost::asio::yield_context& yield,
-    bool repairEnabled = false)
-{
-    clio::LogService::info() << "Beginning migration";
-    auto const ledgerRange = backend.hardFetchLedgerRangeNoThrow(yield);
-
-    /*
-     * Step 0 - If we haven't downloaded the initial ledger yet, just short
-     * circuit.
-     */
-    if (!ledgerRange)
+    void
+    perform()
     {
-        clio::LogService::info() << "There is no data to migrate";
-        return;
+        /*
+         * Step 3 - Drop the old `issuer_nf_tokens` table, which is replaced by
+         * `issuer_nf_tokens_v2`. Normally, we should probably not drop old
+         * tables in migrations, but here it is safe since the old table wasn't
+         * yet being used to serve any data anyway.
+         */
+        clio::LogService::info() << "Running " << tag_;
+
```

Comment:
don't forget to clear the resume state here

Reply:
I did not do it deliberately - don't think it's important as step3 is tiny. If the last saved resume state was at the end of step2 then the resume operation would not take long to write the last page of step2 again and jump over to step3. But i can add it here for completeness 👍